applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_lagrange_dirichlet_condition.cpp

```diff
@@ -0,0 +1,458 @@
+        }
+    }
+
+    if (apply_constraints)
+    {
+        Matrix lagrange_matrix = ZeroMatrix(matrix_size, matrix_size);
+
+         // Loop over Lagrange Multipliers
+        for (unsigned int i = 0; i < number_of_nodes; i++)
+        {
+            // Loop over shape functions of displacements
+            for (unsigned int j = 0; j < number_of_nodes; j++)
+            {
+                const double NN = Variables.N[j] * Variables.N[i];
+                const unsigned int ibase = i * dimension + dimension * number_of_nodes;
+                const unsigned int jbase = j * dimension;
+
+                // Matrix in following shape:
+                // |aug*H   H^T|
+                // |H       0  |
+                // Lambda in X
+                for (unsigned int k = 0; k < dimension; k++)
+                {
+                    lagrange_matrix(jbase+k, jbase+k) = NN * augmentation_factor;
+                    lagrange_matrix(ibase+k, jbase+k) = NN;
+                    lagrange_matrix(jbase+k, ibase+k) = NN;
+                }
+
+
+                //Stabilization needed for some solvers
+                if (stabilization)
+                {
+                    lagrange_matrix(ibase, ibase) = -std::numeric_limits<double>::epsilon();
+                    lagrange_matrix(ibase + 1, ibase + 1) = -std::numeric_limits<double>::epsilon();
+
+                    if (dimension == 3)
+                            lagrange_matrix(ibase +2, ibase +2) = -std::numeric_limits<double>::epsilon();
+
+                }
+            }
+        }
+
+        lagrange_matrix  *= this->GetIntegrationWeight();
+
+        // Calculate LHS Matrix and RHS Vector
+        if ( CalculateStiffnessMatrixFlag == true )
+        {
+            rLeftHandSideMatrix = lagrange_matrix;
+        }
+
+        if ( CalculateResidualVectorFlag == true )
+        {
+            Vector gap_function = ZeroVector(matrix_size);
+            for (unsigned int i = 0; i < number_of_nodes; i++)
+            {
+                const array_1d<double, 3> disp = GetGeometry()[i].FastGetSolutionStepValue(DISPLACEMENT);
+                const int index = dimension * i;
+
+                const array_1d<double, 3> lagrange_multiplier = GetGeometry()[i].FastGetSolutionStepValue(VECTOR_LAGRANGE_MULTIPLIER);
+                const int lagrange_index = dimension * i + dimension * number_of_nodes;
+
+                for (unsigned int j = 0; j < dimension; j++){
+                    gap_function[index+j]          = disp[j] - imposed_displacement[j];
+                    gap_function[lagrange_index+j] = lagrange_multiplier[j];
+                }
+            }
+
+            noalias(rRightHandSideVector) -= prod(lagrange_matrix, gap_function);
+        }
+
+    }
+    else{
+        // To improve stability: use identity matrix to avoid nonzero diagonal LHS matrix
+        if ( CalculateStiffnessMatrixFlag == true )
+        {
+            noalias(rLeftHandSideMatrix) = IdentityMatrix(matrix_size);
```

Comment:
Again, you are indirectly using ALM, but in a not consistent manner (if I understood correctly what have you done)